### ISSUE TITLE
I need a publish method where I can pass in the byte count

### ISSUE THREAD

**Issue by BrannonKing** [nats-io/nats.net.v1#171]
I would like to be able to pass in the byte count that goes with my byte[] passed to the Publish method. In other words, I want to set the "num" value in Connection.publish with a parameter so that I don't have to make an additional array copy of my data after serialization.

**Comment by sixlettervariables** [comment]
Would you also want an offset parameter?

Something like:
```
void Publish(string subject, byte[] data, int count);
void Publish(string subject, byte[] data, int offset, int count);
void Publish(string subject, string reply, byte[] data, int count);
void Publish(string subject, string reply, byte[] data, int offset, int count);
```

**Comment by BrannonKing** [comment]
I don't need "offset" in my current application, but there have been some in times past where it would be handy. Those overloads you listed look fine.

**Comment by ColinSullivan1** [comment]
I think `Publish` APIs with offset and count would keep the API minimized while also handling a wider use case.  For congruence, the `Request*` APIs could provide offsets and count as well.

**Comment by aricart** [comment]
@BrannonKing this is not needed if your 'array' is actually a slice.
If you have a 'slice', you can subrange the slice, the backing of the slice remains to be the original slice. 

So all you need to do is subrange your source slice, and the api will send the subset of bytes, without any data copying. Have a look at this sample:
https://play.golang.org/p/XHs7KlPm78

**Comment by BrannonKing** [comment]
C# doesn't support native array subranges. In order to sub-range a collection in C#, the API would still have to change. If it took a `IReadOnlyList<byte>`, it's easy to build a sub-range wrapper around around that interface. Another option is to take `ArraySegment<byte>`, which bundles the Count and Offset parameters with the array.

**Comment by aricart** [comment]
bah! sorry thought it was go-nats!

**Comment by sixlettervariables** [comment]
vNext of the CLR will have [`Span<T>`](https://github.com/dotnet/coreclr/issues/5851) (rough analog of go slices).

**Comment by BrannonKing** [comment]
@sixlettervariables , I see that you prototyped this. What's the status on a merge?

**Comment by sixlettervariables** [comment]
@BrannonKing I temporarily stepped onto another project at work and just never pushed my final changes. I've opened a PR.

**Comment by ColinSullivan1** [comment]
Resolved by #182.
